File: Question7/Question7.cpp
#include <bits/stdc++.h>
using namespace std;
bool get_cycle_state(int current, vector<int>*transfers,bool *visited,bool *stac)
{
    visited[current] = true;
    stac[current]=true;
    for(auto iter: transfers[current])
    {  if (!visited[iter])
       {if(get_cycle_state(iter,transfers, visited,stac)==true)
            return true;
        }
        else if (stac[iter]==true)
            return true;}
    stac[current]=false;
    return false;
}
bool solution(int n, int l, vector<int>*transfers)
{  
    bool visited[n],stac[n];
    for(int i=0;i<n;i++)
    {
    visited[i]=false;
    stac[i]=false;
    }
   for(int i=0;i<n;i++)
   {
    if (!visited[i])
        {
            if(get_cycle_state(i,transfers,visited,stac)==true)
            return true;
            }
   }
   return false;
}
int main() {
	
	int n, l;
    cin >> n >> l;
  
    vector<int>transfers[n];
    vector<int> transfer(2);
    for (int i = 0; i < l; i++) {
      cin >> transfer[0] >> transfer[1];
      if (transfer[0] != transfer[1])
      transfers[transfer[0]].push_back(transfer[1]);

    }
    bool result = solution(n, l, transfers);
    result ? cout<< "Ineligible" : cout<< "Eligible";
	
	return 0;
}

File: Question7/Question7.py
from typing import List
def get_cycle_state(trans,cur_node, visited,stack):
    visited[cur_node] = True
    stack[cur_node] = True
    for adjacent_node in trans[cur_node]:
        if visited[adjacent_node] == False:
            if get_cycle_state(trans,adjacent_node, visited,stack) == True:
                return True
        elif stack[adjacent_node] == True:
            return True
    stack[cur_node] = False
    return False
def solution(n: int, l: int, transfers: List[List[int]]) -> bool:
    trans={i:[] for i in range(n)}
    for i in range(l):
        if transfers[i][0]!=transfers[i][1]:
            trans[transfers[i][0]].append(transfers[i][1])
            
    visited = [False] * (n)
    stack = [False] * (n)
    for cur_node in range(n):
        if visited[cur_node] == False:
            if get_cycle_state(trans,cur_node,visited,stack) == True:
                return True
    return False
def main():
    n, l = [int(i) for i in input().split(" ")]
    transfers = []
    for i in range(l):
      line = input().split(" ")
      transfers.append([int(line[0]), int(line[1])])
    result = solution(n, l, transfers)
    print("Ineligible" if result else "Eligible")
    
if __name__ == '__main__':
    main()
